cl/opinion_page/utils.py

```diff
@@ -162,3 +133,215 @@ def generate_docket_entries_csv_data(docket_entries):
+        # Filter results by status (e.g., Precedential)
+        # Update URL parameters accordingly
+        search_params[
+            f"stat_{PRECEDENTIAL_STATUS.get_status_value(settings.RELATED_FILTER_BY_STATUS)}"
+        ] = True
+        url_search_params = {
+            f"stat_{PRECEDENTIAL_STATUS.get_status_value(settings.RELATED_FILTER_BY_STATUS)}": "on"
+        }
+
+    if is_bot(request) or not sub_opinion_pks:
+        return [], [], url_search_params, [], 0, False
+
+    cached_citing_results, cached_citing_cluster_count, timeout_cited = (
+        await cache.aget(cache_citing_key) or (None, 0, False)
+    )
+    cached_related_clusters, timeout_related = (
+        await cache.aget(mlt_cache_key) or (None, False)
+        if settings.RELATED_USE_CACHE
+        else (None, False)
+    )
+    # Prepare cited and related cluster queries if not cached results.
+    cluster_search = OpinionClusterDocument.search()
+    cited_search_query = (
+        None
+        if cached_citing_results is not None
+        else await build_cites_clusters_query(cluster_search, sub_opinion_pks)
+    )
+    related_search_query = (
+        None
+        if cached_related_clusters is not None
+        else await build_related_clusters_query(
+            cluster_search, sub_opinion_pks, search_params
+        )
+    )
+    if related_search_query:
+        related_search_query = related_search_query.extra(
+            size=settings.RELATED_COUNT, track_total_hits=False
+        )
+    try:
+        # Execute the MultiSearch request as needed based on available
+        # cached results
+        multi_search = MultiSearch()
+        response_index = 0
+        related_index = citing_index = None
+        if related_search_query:
+            multi_search = multi_search.add(related_search_query)
+            related_index = response_index
+            response_index += 1
+        if cited_search_query:
+            multi_search = multi_search.add(cited_search_query)
+            citing_index = response_index
+        multi_search.params(
+            timeout=f"{settings.ELASTICSEARCH_FAST_QUERIES_TIMEOUT}s"
+        )
+        responses = multi_search.execute() if multi_search._searches else []
+        related_clusters: list[OpinionClusterDocument] = (
+            list(responses[related_index])
+            if related_index is not None
+            else cached_related_clusters or []
+        )
+        citing_clusters: list[OpinionClusterDocument] = (
+            list(responses[citing_index])
+            if citing_index is not None
+            else cached_citing_results or []
+        )
+        citing_cluster_count: int = (
+            responses[citing_index].hits.total.value
+            if citing_index is not None
+            else cached_citing_cluster_count or 0
+        )
+        timeout_related = False if related_clusters else timeout_related
+        timeout_cited = False if citing_clusters else timeout_cited
```

Comment:
We could potentially simplify the code within the try block by moving some logic outside of it. For example, we might be able to compute `related_clusters`, `citing_clusters`, `citing_cluster_count`, and `timeouts` after the try statement (using a `dataclass` could be helpful for this).

Reply:
Yeah, using the dataclass helped to also simplify the code here. Refactor applied.